Feature/issue#51 - 친구 기능 개발

minju412 opened the pull request:
### ✍️ Description
- "친구 관계인 회원들끼리 플레이리스트를 공유할 수 있다"는 비즈니스 요구사항에 따라 친구 관계를 설정할 수 있도록 합니다.

<br>

### 🎯 Key Changes
- `Member`에서 복합키를 제거했습니다.
    - 이유는, 친구 요청 시 Member의 id가 필요한데 복합키를 사용하면 로직이 복잡해지기 때문입니다.
- #51 에 따라 아래 기능을 추가했습니다.
    - 친구 요청
    - 친구 요청 수락
    - 내가 친구 요청 보낸 회원 목록 조회
    - 나에게 친구 요청 보낸 회원 목록 조회
    - 나와 서로 친구인 회원 목록 조회
- 정방향 레코드와 역방향 레코드를 모두 저장합니다.
    - 이유는, 정방향 레코드만 저장하면 친구인 상태에서 역방향 요청이 성공한다는 문제가 있기 때문입니다.

<br>

### 💬 To Reviewers
- `Member` 엔티티가 멤버변수로 `List<Friendship>`를 가지고 있어야 할 지 고민입니다.
    -  `List<Friendship>` 없이도 repository에서 조회할 수 있지만 효율성 측면에서 가지고 있는 것이 좋을지 궁금합니다!

minju412 committed:
refactor: setAuthentication()의 파라미터로 OAuth2AuthenticationToken 전달

- OAuth2AuthenticationToken은 OAuth 2.0을 사용하여 인증된 사용자를 나타내는 Authentication 구현체입니다.
- 따라서, OAuth를 통해 로그인하는 상황에 맞도록 OAuth2AuthenticationToken을 SecurityContext에 저장합니다.

minju412 committed:
refactor(Member): 복합키 제거

- provider와 providerId를 복합키로 사용하는 로직을 제거했습니다.
- 기존 id 대신 복합키를 사용한 이유는 id가 식별자로서 기능을 하지 않았기 때문입니다.
- 하지만 '친구 신청' 시 Long 타입의 id를 사용하는 것이 훨씬 간편하기 때문에 복합키를 제거했습니다.

minju412 committed:
feat: CustomException 추가

- null을 반환하는 대신 Exception을 발생시킵니다.
- BusinessException 로직이 코드에 반영된 이후에는 BusinessException으로 변경할 예정입니다.

minju412 committed:
feat: 친구 신청 api 추가

- 친구 관계(Friendship) 엔티티를 추가했습니다.
- 친구 관계 테이블은 친구를 신청한 회원(fromMember)와 친구 요청을 받은 회원(toMember) 정보를 담고 있습니다.
- 회원:회원 관계를 N:N으로 설정하였고 중간 테이블로 Friendship을 사용합니다.
- 나와 서로 친구인 회원 조회 기능을 위해 정방향 레코드 + 역방향 레코드 저장

minju412 committed:
refactor: 패키지 이름 변경

- oauth 사용자 정보가 담긴 패키지들로 이루어져 있어 userinfo로 변경

minju412 committed:
feat: 내가 친구 요청 보낸 회원 조회 api 추가

- fromMember는 친구 요청을 보낸 회원이고 toMember는 친구 요청을 받은 회원입니다.
- Friendship 레코드 중 fromMember가 로그인한 사용자이면서 isFriend가 False인 회원을 조회합니다.

minju412 committed:
feat: 나에게 친구 요청 보낸 회원 조회 api 추가

- Friendship 레코드 중 toMember가 로그인한 사용자이면서 isFriend가 FALSE인 회원을 조회합니다.

minju412 committed:
feat: 나와 서로 친구인 회원 조회 기능 api 추가

- 셀프조인을 사용하여 fromMember와 toMember가 양방향 관계이면서 isFriend가 TRUE임을 확인합니다.
- 친구 요청 시 정방향 레코드의 isFriend는 TRUE로, 역방향 레코드의 isFriend는 FALSE로 수정했습니다.
    - 이유는, 내가 친구 요청 보낸 회원과 나에게 친구 요청 보낸 회원을 구분하기 위해서 from과 to의 구분이 필요한데,
    - isFriend를 그 구분값으로 사용합니다.

minju412 committed:
feat: 친구 요청 수락 api 추가

- 친구 요청을 보낸 회원의 id로 Friendship을 조회합니다.
- 이후 해당 Friendship의 isFriend를 True로 갱신하여 친구 관계를 성립합니다.
- 해당 Friendship이 존재하지 않는다면 에러를 발생시킵니다.

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/auth/oauth/CustomOauth2UserService.java:
```diff
@@ -41,30 +40,30 @@ private OAuth2UserInfo getOAuth2UserInfo(OAuth2UserRequest userRequest, OAuth2Us
+			.providerId(oAuth2UserInfo.getProviderId())
 			.build();
 		return memberRepository.save(member);
 	}
```
로직을 메서드로 묶어서 잘 나누어주셨네요! 그런데 `getExistingMember()`메서드의 경우 `null` 을 반환하고 있는데 이런 경우에는 사이드 이펙트에 노출될 수 있는 가능성이 높아집니다! `Optional`을 사용하는건 어떨까요? 하지 않고 코드를 짜보면 좋을 것 같아요.

`tip` : `orElseGet()` 에 대해서 찾아보세요!

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/auth/oauth/OAuth2Provider.java:
```diff
@@ -14,4 +18,11 @@ public enum OAuth2Provider {
 	private final String providerName;
 	private final String attributeKey;
+
+	public static OAuth2Provider getEnum(String providerName) {
+		return Arrays.stream(OAuth2Provider.values())
+			.filter(value -> value.providerName.equals(providerName))
+			.findFirst()
+			.orElseThrow(() -> new NotRegisteredProviderException());
+	}
```
코드의 문맥 상 더 좋은 메서드 명이 있을 것 같아요!

minju412 replied:
더 구체적으로 `fromPorivderName()`으로 변경하려 했으나, 사용되지 않는 메서드라 제거했습니다! 😂

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/friendship/repository/FriendshipRepository.java:
```diff
@@ -0,0 +1,29 @@
+	List<Friendship> findFriendshipsByFromMemberAndIsFriend(Member fromMember, Boolean isFriend);
+	List<Friendship> findFriendshipsByToMemberAndIsFriend(Member toMember, Boolean isFriend);
+	@Query(
+		"SELECT f.toMember FROM Friendship f " +
+			"JOIN Friendship sf " +
+			"ON f.toMember.memberId = sf.fromMember.memberId " +
+			"WHERE f.fromMember = :loginMember " +
+			"AND f.fromMember = sf.toMember " +
+			"AND f.toMember = sf.fromMember " +
+			"AND f.isFriend = true " +
+			"AND sf.isFriend = true"
```
쿼리에서 조건문을 몇개 줄일 수 있을 것 같아요!

minju412 replied:
`ON`절과 중복되는 조건을 제거했습니다!

Nine-JH commented on src/main/java/com/petitapetit/miml/domain/friendship/service/FriendshipService.java:
```diff
@@ -0,0 +1,109 @@
+		Member toMember = memberRepository.findById(createRequest.getToMemberId())
+			.orElseThrow(() -> new MemberNotFoundException());
+		// 정방향 레코드 추가
+		friendshipRepository.save(Friendship.builder()
+			.toMember(toMember)
+			.fromMember(fromMember)
+			.isFriend(Boolean.TRUE)
+			.build());
+		// 역방향 레코드 추가
+		friendshipRepository.save(Friendship.builder()
+			.toMember(fromMember)
+			.fromMember(toMember)
+			.isFriend(Boolean.FALSE)
+			.build());
+	}
```
역방향 대한 아이디어는 매우 좋습니다!
그런데 이를 한번에 저장할 수는 없을까요? JPA에서는 이미 해당 기능을 제공하고 있답니다 😄

Nine-JH reviewed (commented):
고생하셨습니다! 몇몇 코멘트를 남겼으니 한번 고민해보세요!

Jeonwoojae reviewed (approved):
고생하셨습니다👍

minju412 committed:
refactor(CustomOAuth2UserService): null을 반환하는 코드 제거

- 회원 정보가 없는 경우, null을 반환하지 않고 orElseGet()을 이용하여 새로운 회원을 생성합니다.
- 그 이유는 null을 반환하는 경우 사이드 이펙트에 노출될 수 있는 가능성이 높아지기 때문입니다.

minju412 committed:
refactor(CustomOAuth2UserService): 클래스 이름 변경

- CustomOauth2UserService -> CustomOAuth2UserService 로 변경했습니다.

minju412 committed:
refactor: 사용되지 않는 메서드 및 필드 제거

- OAuth2Provider Enum 내 toEnum() 메서드 제거
- CustomOAuth2User 클래스 내 provderName 필드 제거

minju412 committed:
refactor(OAuth2AuthenticationSuccessHandler): 클래스 이름 변경

- CustomOAuth2AuthenticationSuccessHandler -> OAuth2AuthenticationSuccessHandler 로 변경했습니다.

minju412 committed:
refactor(FriendshipRepository): findFriends 쿼리 수정

- ON 절과 중복되는 조건을 제거했습니다.

minju412 committed:
refactor(FriendshipService): saveAll() 메서드로 db 접근 횟수 감소

- save() 메서드를 두 번 호출한 코드를 saveAll() 메서드로 변경했습니다.
- 이로써 db 접근 횟수를 줄일 수 있습니다.
